server/src/main/java/io/spine/examples/shareaware/server/market/MarketDataProvider.java

```diff
@@ -0,0 +1,132 @@
+
+import io.spine.core.UserId;
+import io.spine.examples.shareaware.Share;
+import io.spine.examples.shareaware.market.event.MarketSharesUpdated;
+import io.spine.server.integration.ThirdPartyContext;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.google.common.util.concurrent.Uninterruptibles.sleepUninterruptibly;
+import static java.util.concurrent.Executors.newSingleThreadExecutor;
+
+/**
+ * Provides data about currently available shares on the market to the ShareAware context.
+ */
+public final class MarketDataProvider {
+
+    private static MarketDataProvider instance = null;
+
+    /**
+     * The name of the Bounded Context on behalf of which the data
+     * about available shares on the market will be provided.
+     */
+    private static final String tenantName = "MarketData";
+
+    /**
+     * A single-tenant instance of the {@code ThirdPartyContext}
+     * that represents {@value tenantName} as a Bounded Context to the ShareAware application.
+     */
+    private final ThirdPartyContext marketContext =
+            ThirdPartyContext.singleTenant(tenantName);
+
+    /**
+     * The ID of the {@value tenantName} Bounded Context.
+     *
+     * <p>Provides the information about who emitted the {@code MarketSharesUpdated} event.
+     */
+    private final UserId actor = UserId
+            .newBuilder()
+            .setValue(tenantName)
+            .vBuild();
+
+    /**
+     * The thread executor that allows {@code MarketDataProvider} to work
+     * in a separate thread from the ShareAware application.
+     */
+    private final ExecutorService marketThread = newSingleThreadExecutor();
+
+    /**
+     * Controls the thread in which {@code MarketDataProvider}
+     * is executed to be active or not.
+     */
+    private final AtomicBoolean active = new AtomicBoolean();
+
+    /**
+     * Prevents instantiation of this class.
+     */
+    private MarketDataProvider() {
+    }
+
+    /**
+     * Creates the instance of the {@code MarketDataProvider} if there is no such,
+     * otherwise returns the existing.
+     */
+    public static synchronized MarketDataProvider instance() {
+        if (instance == null) {
+            instance = new MarketDataProvider();
+        }
+        return instance;
+    }
+
+    /**
+     * Emits the {@code MarketSharesUpdated} event with a specified periodicity
+     * on behalf of the {@value tenantName} Bounded Context.
+     */
+    synchronized void runWith(Duration period) {
+        active.set(true);
+        marketThread.execute(() -> {
+            while (active.get()) {
+                sleepUninterruptibly(period);
+                emitEvent();
+            }
+        });
+    }
+
+    /**
+     * Stops the event emission.
+     */
+    synchronized void stop() {
+        active.set(false);
+        marketThread.shutdown();
```

Comment:
I think, it should probably be `shutdownNow`.

Also, let's explicitly what happens if we want to start the provider again, after stopping it. Is it allowed? Please expand the class-level documentation on this.